### PR TITLE
Add CI jobs

### PR DESCRIPTION
**Pull request by cottsay**
TODO list for this feature:
- [x] Workspace bringup
- [x] Colcon build and test
- [x] Docker invalidation based on
  - Changes in the repos file
  - Changes in the rosdep index
  - Changes in the base OS' package versions
- [x] Output artifact for use as underlay for subsequent job
- [x] ccache
- [ ] Jenkins parameters
  - Namely, additional repos URLs, target branch, build options, etc
- [ ] Lots of other `ros_buildfarm_config` values (priority, etc)
- [ ] Scheduling
- [ ] Overlay job
  - Need to determine how to target a specific package or packages

### CODE REVIEW

**Comment by cottsay** [comment]
Basic workflow:
- Generate, build and run `ci_task_generation` in docker, which generates `ci_create_workspace` and `ci_colcon_build`.
- Build and run `ci_create_workspace`, which pulls the repos files and packages, as well as enumerates the dependencies needed to build and run tests
- Copy `install_list.txt` into docker directory, build and run `ci_colcon_build`

File: ros_buildfarm/config/index.py [hunk 1]
```diff
@@ -87,6 +88,13 @@ def __init__(self, data, base_url):
                 for key in unset_keys:
                     self.distributions[distro_name][key] = value_types[key]()
 
+        self.ci_builds = {}
+        if 'ci_builds' in data and data['ci_builds']:
+            assert isinstance(data['ci_builds'], dict)
+            for k, v in data['ci_builds'].items():
+                v = _resolve_url(base_url, v)
+                self.doc_builds[k] = v
```
Comment (nuclearsandwich) [inline ros_buildfarm/config/index.py]: Is this a copy paste error? Should it be?

```suggestion
                self.ci_builds[k] = v
```

**Comment by cottsay** [comment]
Example `ros_buildfarm_config` definition: https://github.com/cottsay/ros_buildfarm_config/blob/cottsay-testfarm/bouncy/ci-build.yaml

**Comment by cottsay** [comment]
Example job: http://34.238.124.21/job/Bci__ubuntu_bionic_amd64/16/

**Comment by dirk-thomas** [comment]
I took a first peek at this patch and if you are ok with it I would prefer getting #577 landed before. Rebasing this patch and using some of the modified API provided by the other PR should be easier than the other way around and will also make sure this feature works for both `catkin_make_isolated` as well as `colcon`.

**Comment by cottsay** [comment]
I had the same conclusion, @dirk-thomas. I'll work on getting something running for the overlay jobs in the mean time.

File: ros_buildfarm/templates/ci/ci_job.xml.em [hunk 2]
```diff
@@ -241,7 +241,10 @@ parameters = [
 ))@
 @(SNIPPET(
     'archive_artifacts',
-    artifacts=['install_merged.tar.gz'],
+    artifacts=[
+      'install_merged.tar.gz',
+      'docker_colcon_build/install_list.txt',
```
Comment (nuclearsandwich) [inline ros_buildfarm/templates/ci/ci_job.xml.em]: Not saying we shouldn't but could we document the utility of this?

Comment (cottsay) [reply ros_buildfarm/templates/ci/ci_job.xml.em]: I added the `install_list.txt` while whiteboarding the overlay jobs. At this point, I think it would be better to re-query rosdep after downloading the merged install archive in order to satisfy runtime and downstream dependencies. It would be great to find a way to exclude the build-time dependencies taken by the underlay when running overlay jobs, but I don't think there is a straightforward way to do that right now. Could be a future optimization.

tl;dr - I'll remove the `install_list.txt` because I don't think we'll need it.

File: ros_buildfarm/colcon_workspace.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    if parent_result_spaces is not None:
+        for parent_result_space in reversed(parent_result_spaces):
+            setup_file = os.path.join(parent_result_space, 'setup.sh')
+            if os.path.exists(setup_file):
+                cmd = 'COLCON_CURRENT_PREFIX=%s . %s && %s' % (parent_result_space, setup_file, cmd)
+
+    print("Invoking '%s' in '%s'" % (cmd, workspace_root))
+    return subprocess.call(
+        cmd, cwd=workspace_root, shell=True, stderr=subprocess.STDOUT,
+        env=env, executable='/bin/bash')
```
Comment (dirk-thomas) [inline ros_buildfarm/colcon_workspace.py]: This file should be obsolete.

File: ros_buildfarm/config/ci_build_file.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            "Expected file type is '%s', not '%s'" % \
+            (CIBuildFile._type, data['type'])
+
+        assert 'version' in data, \
+            "Source build file for '%s' lacks required version information" % \
+            self.name
+        assert int(data['version']) in [1], \
+            ("Unable to handle '%s' format version '%d', please update " +
+             "rosdistro (e.g. on Ubuntu/Debian use: sudo apt update && " +
+             "sudo apt install --only-upgrade python-rosdistro)") % \
```
Comment (dirk-thomas) [inline ros_buildfarm/config/ci_build_file.py]: The version of the build file has nothing to do with `rosdistro`.

File: ros_buildfarm/templates/ci/ci_job.xml.em [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+  <scmCheckoutRetryCount>2</scmCheckoutRetryCount>
+  <assignedNode>@(node_label)</assignedNode>
+  <canRoam>false</canRoam>
+  <disabled>@('true' if disabled else 'false')</disabled>
+  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
+  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
+  <triggers>
+@(SNIPPET(
+    'trigger_timer',
+    spec='0 23 * * *',
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/ci/ci_job.xml.em]: It would be good to configure the timer trigger in the build file.

**Comment by cottsay** [comment]
I'm going to re-open this PR against master now that the colcon stuff has been merged, so I'll close it here.
